### ISSUE TITLE
Apache loading sqlsrv.so crashes with error invalid free(). build with -Ggdb -O0 is a workaround

### ISSUE THREAD

**Issue by tombay** [microsoft/msphpsql#1302]
.+Please check the [FAQ (frequently-asked questions)](https://github.com/Microsoft/msphpsql/wiki/FAQ) first. If you have other questions or something to report, please address the following:

 +## PHP Driver version or file name
 +pdo_sqlsrv 5.9.0   stable and pdo_sqlsrv 5.9.0   stable 
 +## SQL Server version
 +Microsoft SQL Server 2019 (RTM) - 15.0.2000.5 (X64) Sep 24 2019 13:48:23 on WIndows10
 +## Client operating system
 +Linux  5.4.0-84-generic #94-Ubuntu SMP Thu Aug 26 20:27:37 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
 +## PHP version
 +PHP 8.0.10 (cli) (built: Aug 26 2021 15:50:07) ( NTS ) 
 +## Microsoft ODBC Driver version
 +Driver={ODBC Driver 17 for SQL Server};  (libmsodbcsql-17.8.so.1.1)
 +## Table schema
 +
 +## Problem description
 +Running Ubuntu Apache Webserver (Server version: Apache/2.4.41 (Ubuntu)) with /usr/lib/apache2/modules/libphp8.0.so
loaded. Trying to add both (or either) pdo_sqlsrv.so or sqlsrv.so via enablin modules in the php ini. 
 +## Expected behavior and actual behavior
 + I expected the driver to load seamlessly - but the apache server crashes. I have supplied a backtrace in attached file.
 +## Repro code or steps to reproduce
 sudo pecl install sqlsrv ; sudo service apache2 restart

I rebuilt the driver with CFLAGS and CPPFLAGS set to -g -O0 and it crashed.
I then rebuilt the driver with CFLAGS and CPPFLAGS set to -ggdb -O0 and it is now working. 
in analysing the crash crash_o0_g gdb states warning: .dynamic section for "/usr/lib/php/20200930/sqlsrv.so" is not at the expected address (wrong library or version mismatch?), I am not sure if this is because of unoptimised code (I suspect not) but maybe it is indicating my problem. This is not a production system I am working on, and so I can work on my system with the debug build. 
I have attached a testsqlsrv.php which runs from the command line and so the issue is not with php module setup. It is rather to do with the integration with apache.
[sqlsrv_O0_g_crash.txt](https://github.com/microsoft/msphpsql/files/7168099/sqlsrv_O0_g_crash.txt)
[sqlsrv_O2_g_crash.txt](https://github.com/microsoft/msphpsql/files/7168102/sqlsrv_O2_g_crash.txt)
[sqlsrv_environ.txt](https://github.com/microsoft/msphpsql/files/7168104/sqlsrv_environ.txt)

**Comment by tombay** [comment]
Attached is the gcc version.
An observation which I don't know how to address is outlined by % file /usr/lib/php/20200930/*.so
...
/usr/lib/php/20200930/pdo.so:        ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, BuildID[sha1]=08cd37724e59464b8ba8ac45b78abd02deea1ce1, stripped
/usr/lib/php/20200930/pdo_sqlsrv.so: ELF 64-bit LSB shared object, x86-64, version 1 (GNU/Linux), dynamically linked, BuildID[sha1]=20fd15438f6dacabf4012a8d3eedbe7cee19da89, with debug_info, not stripped
/usr/lib/php/20200930/pgsql.so:      ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, BuildID[sha1]=bbac5e092887db2694e1a205bd047309cee30a7d, stripped
/.
The only .so file with (GNU/Linux) are slqsrv.so and pdo_sqlsrv.so. maybe it explains the issue ? 


/tmp$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/9/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none:hsa
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 9.3.0-17ubuntu1~20.04' --with-bugurl=file:///usr/share/doc/gcc-9/README.Bugs --enable-languages=c,ada,c++,go,brig,d,fortran,objc,obj-c++,gm2 --prefix=/usr --with-gcc-major-version-only --program-suffix=-9 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-plugin --enable-default-pie --with-system-zlib --with-target-system-zlib=auto --enable-objc-gc=auto --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-offload-targets=nvptx-none=/build/gcc-9-HskZEa/gcc-9-9.3.0/debian/tmp-nvptx/usr,hsa --without-cuda-driver --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 9.3.0 (Ubuntu 9.3.0-17ubuntu1~20.04)

**Comment by yitam** [comment]
Hi @tombay , which ubuntu are you using? 

If I understand you correctly, you were trying to debug the drivers? In that case, you might need to build PHP from source with the `--enable-debug` option. Without altering the flags, try this first. If necessary, modify the config.m4 files to use `-O0` instead of `-O2` to avoid compiler optimizations.

Will also do some investigation and get back to you.

**Comment by tombay** [comment]
Hi Jenny.
Ubuntu 20.04 (lts).
I am running with -O0 -ggdb. I am only running like that so that the driver loads and Apache keeps running. I was just suspicious that it is a memory allocation issue - interaction between stdc and c++ . 
I'd someone could suggest how to have the image link as SYSV I would do that. I had a cursory look but could see no ld flags to do that. I am quite happy for someone else to debug the driver, I am glad it loads so I can develop my self app. It works and so it's probably a build environment issue I can't resolve %-?

**Comment by tombay** [comment]
Woops

**Comment by tombay** [comment]
I didn't mean to close this. 😕

**Comment by yitam** [comment]
hi @tombay , it's likely something to do with your settings or env in Ubuntu 20.04. I tested on both Ubuntu 20.04 and Ubuntu 18.04 (they are relatively clean) and found no issues with Apache. The images below are cropped from Ubuntu 20.04:

![image](https://user-images.githubusercontent.com/21340756/133858450-6151820c-f078-4a2c-885e-4308ce40112e.png)

![image](https://user-images.githubusercontent.com/21340756/133858495-64c90b49-363e-49db-a66c-c445dc29b0fb.png)

How did you install PHP or Apache? Did you follow [our instructions](https://github.com/microsoft/msphpsql/blob/master/Linux-mac-install.md#installing-the-drivers-on-ubuntu)? Or did you build from source?

**Comment by tombay** [comment]
Hi @yitam .
Thanks for checking this out. 
Apache came from the Ubuntu archive
My php was installed from the ondrej repositories.
I spent some time trying to ascertain what was causing the crashes.
With CFLAGS and CXXFLAGS set to -ggdb -O0 the shared images load.
With CFLAGS and CXXFLAGS set to -ggdb -O2 the shared images crash.
I will close this defect as it does seem environmental issue. My development task was involving a php framework and pgsql,mysql,sqlsrv. Now that I have sqlsrv working I can get back to it. Out of interest I built Apache/2.5.1-dev and the images worked. But ubuntu 20.04 (my world) standard Apache is the 2.4 stream %-)
Thanks again for your assistance

Fyi
$ grep -i php /etc/apt/sources.list.d/*|grep focal
/etc/apt/sources.list.d/ondrej-ubuntu-php-focal.list:deb http://ppa.launchpad.net/ondrej/php/ubuntu focal main
/etc/apt/sources.list.d/ondrej-ubuntu-php-focal.list:# deb-src http://ppa.launchpad.net/ondrej/php/ubuntu focal main
Apache came from the Ubuntu archive
$ apt-cache showpkg apache2
Package: apache2
Versions: 
2.4.41-4ubuntu3.4 (/var/lib/apt/lists/archive.ubuntu.com_ubuntu_dists_focal-updates_main_binary-amd64_Packages) (/var/lib/dpkg/status)
 Description Language: 
                 File: /var/lib/apt/lists/archive.ubuntu.com_ubuntu_dists_focal_main_binary-amd64_Packages
